### ISSUE TITLE
giraffe alignment

### ISSUE THREAD

**Issue by userzxyz** [vgteam/vg#3316]
Hello,

I have used `giraffe` for alignment using:
```
vg autoindex --workflow giraffe -g chr1.gfa -p pangenome
vg giraffe -H pangenome.giraffe.gbwt -g pangenome.gg -mpangenome.min -d pangenome.dist -f sample_R1_paired.fastq.gz -f sample_R2_paired.fastq.gz >mapped.gam
vg stats -a mapped.gam

```

With the stats, I got:
```
Total alignments: 33850102
Total primary: 33850102
Total secondary: 0
Total aligned: 275374
Total perfect: 165786
Total gapless (softclips allowed): 261643
Total paired: 33850102
Total properly paired: 18
Insertions: 18382 bp in 8489 read events
Deletions: 25545 bp in 9845 read events
Substitutions: 492971 bp in 492971 read events
Softclips: 2098177 bp in 52728 read events
```

My concern is the `Total properly paired: 18 `. Can it be that low because I used `gfa` for only one chromosome? 

Additionally, I tried finding how to use this alignment file for downstream analysis specifically variant calling. For `vg map`, I used:
```
vg surject -x chr1.xg -b chr1.gam > chr1.bam
vg pack -x chr1.xg -g chr1.gam -Q 5 -o aln.pack
vg call chr1.xg -k aln.pack > chr1_calls.vcf
```
I could not find the documentation for something similar with `giraffe` output. Thank you!

**Comment by jeizenga** [comment]
If you used a graph for one chromosome with raw shotgun sequencing data, you can get a lot of weird behavior because the majority of reads will have nowhere good to map. My first guess would be that Giraffe is not learning a good fragment length distribution because of the interchromosomal mappings, which is affecting its ability to identify properly paired reads.

For variant calling, you can use a GAM from Giraffe the same way as a GAM from VG map.

**Comment by jeizenga** [comment]
Most of our graph formats label the path of the reference sequence, which is required for `vg surject` so that it can identify the mapped location on the reference. It seems that the gbwt/gg pipeline doesn't preserve that information (GBWT path storage is a bit different since it's actually storing a large collection of paths).

One option would be to make the XG directly from the GFA file using `vg convert`. @jltsiren are there any pitfalls to avoid with making the XG and GBWTGraph separately? It looked like we might have run into some problems in https://github.com/vgteam/vg/issues/3315.

**Comment by userzxyz** [comment]
@jeizenga Thank you! I tried:
```
vg convert -g chr1.gfa -x pangenome.gg > pangenome_1.xg
vg surject -x pangenome_1.xg -b mapped.gam > mapped.bam
vg pack -x pangenome_1.xg -g mapped.gam -Q 5 -o aln.pack
vg call pangenome_1.xg -k aln.pack > mapped.vcf
```

However, I got this warning for `surject` command line:
`warning[vg::Surjector]: Refusing to perform very large alignment against 160412 bp strand split subgraph for read sample.169050; suppressing further warnings.
`

**Comment by jltsiren** [comment]
I'm not sure what `vg autoindex` actually does if you give it GFA input and want Giraffe indexes. I believe GBWTGraph construction and XG construction will interpret GFA segments in the same way if no segment is longer than 1024 bp and all segments have either numerical or non-numerical names. If there are long segments or both numerical and non-numerical names, there will be many subtle edge cases. In general, I would recommend parsing the GFA file only once and using the results as the ground truth.

The ideal situation is that you have a GFA file with reference paths as P-lines and haplotype paths as W-lines. Then you can import the GFA with `vg gbwt` and build an XG with reference paths with `vg convert`:

```
vg gbwt -o graph.gbwt -g graph.gg -G graph.gfa
vg convert -x -b graph.gbwt graph.gg > graph.xg
# And then find snarls and build distance index from the XG
```

If you have both reference paths and haplotypes as P-lines, you must first ensure that all paths use the same naming scheme. For example, `sample_contig_haplotype`, where `sample` and `contig` are strings and `haplotype` is an integer. Reference paths should use `ref` as the sample name and have distinct contig names.

```
vg gbwt -o graph.gbwt -g graph.gg --path-regex "(.*)_(.*)_(.*)" --path-fields _SCH -G graph.gfa
vg convert -x -b graph.gbwt --ref-sample ref graph.gg > graph.xg
```

**Comment by userzxyz** [comment]
Thank you, @jltsiren! I tried using:
```
vg gbwt -o pangenome.gbwt -g pangenome.gg -G chr1.gfa
vg convert -x -b pangenome.gbwt pangenome.gg > graph.xg
vg snarls graph.xg > graph.snarls
vg index graph.xg -s graph.snarls -j graph.dist
```
However, the index job got killed without any error.

**Comment by jltsiren** [comment]
You should use option `-T` to include trivial snarls in the `vg snarls` command. I'm not sure if that will fix the particular crash you had, though.

**Comment by userzxyz** [comment]
Thank you! Using -T solved the index problem. Then I used:
```
vg mod -X 1000 smooth.gfa > graph.fixed.gfa 
vg gbwt -o sample.gbwt -g sample.gg -G graph.fixed.gfa 
vg convert -x -b sample.gbwt sample.gg > sample.xg 
vg snarls -T sample.xg > sample_1.snarls
vg index sample.xg -s sample_1.snarls -j sample_1.dist
vg giraffe -f sample_R1_paired.fastq.gz -f sample_R2_paired.fastq.gz -x sample.xg -H sample.gbwt > reads.gam
```
I got `Segmentation fault`. I found another issue with same problem and an updated version solved it. I am using vg/1.32.0 and I will try vg/1.33.0 and see how it goes.

**Comment by jltsiren** [comment]
This could be usability features leading to undesired behavior.

Giraffe uses four indexes for mapping: GBWT (`-H graph.gbwt`), GBWTGraph (`-g graph.gg`), distance index (`-d graph.dist`), and minimizer index (`-m graph.min`). If you do not provide all of them with command line arguments, Giraffe guesses the base name from the name of the graph (`-x graph.xg` or `-g graph.gg`) and assumes that files with the same base name and the correct extension contain the missing indexes. If that fails, Giraffe tries to build the missing indexes from the ones it can find.

In your case, any of these could be the problem:

* There is an earlier minimizer index `sample.min` corresponding to a different graph. Delete it.
* There is an earlier distance index `sample.dist` corresponding to a different graph. Delete it, and because you built a distance index with another base name, tell Giraffe about it with `-d sample_1.dist`.
* There is an earlier GBWTGraph `sample.gg` corresponding to a different graph. Delete it, as it overrides the provided `sample.xg`.

After all this, please rerun Giraffe with option `-p`. If it still crashes, please post the progress information it prints.

**Comment by userzxyz** [comment]
Thank you @jltsiren! I deleted all those files with same base names and ran the giraffe alignment. However, I still got this:

```
Checking for availability of gg index: Needs to be built. Checking dependencies...
Checking for availability of vg index: Loadable from sample.xg
Checking for availability of gbwt index: Loadable from sample.gbwt
Checking for availability of gbwt index: Loadable from sample.gbwt
Checking for availability of min index: Needs to be built. Checking dependencies...
Checking for availability of gg index: Needs to be built. Checking dependencies...
Checking for availability of vg index: Loadable from sample.xg
Checking for availability of gbwt index: Loadable from sample.gbwt
Checking for availability of dist index: Needs to be built. Checking dependencies...
Checking for availability of vg index: Loadable from sample.xg
Checking for availability of snarls index: Needs to be built. Checking dependencies...
Checking for availability of vg index: Loadable from sample.xg
Building min
Building gg
Loading vg from sample.xg
Loading gbwt from sample.gbwt
Building dist
Building snarls
Segmentation fault
```

**Comment by jltsiren** [comment]
The graph should be at least as "large" as the read set. If you have whole-genome reads, you should map them to a whole-genome graph. Giraffe has not been designed for mapping whole-genome reads to a single-chromosome graph, and nobody really understands what will happen if you try it.

**Comment by userzxyz** [comment]
Thank you, @jltsiren. To merge the chromosome wise graphs, should I merge the .`gfa` output files or the index files. Can you suggest what would be a good way to merge chromosome-wise graph outputs to use for `giraffe`.

**Comment by jltsiren** [comment]
You should merge the GFA files, because some indexes cannot be merged.

Index construction for large GFA files may be slow and/or require a lot of memory, as the construction process has not been optimized for such inputs yet.

**Comment by userzxyz** [comment]
Thank you @jltsiren! I tried finding but could not find any method. Can you please suggest how to merge GFA files. Thank you!

**Comment by AndreaGuarracino** [comment]
@userzxyz, an alternative is [odgi squeeze](https://odgi.readthedocs.io/en/latest/rst/commands/odgi_squeeze.html). You need to convert the GFAs to ODGI format (with `odgi build`), squeeze them in a single ODGI file, and then reconvert the squeezed ODGI file in GFA (with `odgi view`).

**Comment by glennhickey** [comment]
You can also use `vg combine 1.gfa 2.gfa 3.gfa ... > combined.gfa`

**Comment by faithokamoto** [comment]
Seems this was answered at the time.
